kratos/geometries/integration_point_surface_3d.h

```diff
@@ -0,0 +1,473 @@
+     * This type used to hold geometry's points.
+     */
+    typedef  typename BaseType::PointsArrayType PointsArrayType;
+
+    /**
+     * This type used for representing an integration point in
+     * geometry. This integration point is a point with an
+     * additional weight component.
+     */
+    typedef typename BaseType::IntegrationPointType IntegrationPointType;
+
+    /**
+     * A Vector of IntegrationPointType which used to hold
+     * integration points related to an integration
+     * method. IntegrationPointSurface3ds functions used this type to return
+     * their results.
+     */
+    typedef typename BaseType::IntegrationPointsArrayType IntegrationPointsArrayType;
+
+    /**
+     * A Vector of IntegrationPointsArrayType which used to hold
+     * integration points related to different integration method
+     * implemented in geometry.
+     */
+    typedef typename BaseType::IntegrationPointsContainerType IntegrationPointsContainerType;
+
+    /**
+     * A third order tensor used as shape functions' values
+     * container.
+     */
+    typedef typename BaseType::ShapeFunctionsValuesContainerType
+    ShapeFunctionsValuesContainerType;
+
+    /**
+             * A fourth order tensor used as shape functions' local
+             * gradients container in geometry.
+     */
+    typedef typename BaseType::ShapeFunctionsLocalGradientsContainerType
+    ShapeFunctionsLocalGradientsContainerType;
+
+    /**
+             * A third order tensor to hold jacobian matrices evaluated at
+             * integration points. Jacobian and InverseOfJacobian functions
+             * return this type as their result.
+     */
+    typedef typename BaseType::JacobiansType JacobiansType;
+
+    /**
+     * A third order tensor to hold shape functions' local
+     * gradients. ShapefunctionsLocalGradients function return this
+     * type as its result.
+     */
+    typedef typename BaseType::ShapeFunctionsGradientsType ShapeFunctionsGradientsType;
+
+    /**
+     * A third order tensor to hold shape functions' local second derivatives.
+     * ShapefunctionsLocalGradients function return this
+     * type as its result.
+     */
+    typedef typename BaseType::ShapeFunctionsSecondDerivativesType
+        ShapeFunctionsSecondDerivativesType;
+
+
+    /**
+    * A fourth order tensor to hold shape functions' local third derivatives.
+    * ShapefunctionsLocalGradients function return this
+    * type as its result.
+     */
+    typedef typename BaseType::ShapeFunctionsThirdDerivativesType
+        ShapeFunctionsThirdDerivativesType;
+
+    /**
+    * Type of the normal vector used for normal to edges in geomety.
+     */
+    typedef typename BaseType::NormalType NormalType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    IntegrationPointSurface3d(
+        const PointsArrayType& ThisPoints,
+        const CoordinatesArrayType& rLocalCoordinates,
+        const double& rIntegrationWeight,
+        Matrix& rShapeFunctions)
```

Comment:
thanks for your issues. I know that this is a bit difficult and one would have to think precisely. But like this one can pass as many derivatives as wished. In high order this might be very important.

Maybe we can also bypass this with a nested vector.

Reply:
A comment about the array dimensions: Through the standard construction they might have the same problem. The integration points, shape functions and derivatives are all computed in separate functions.

Maybe in the setters one could leave some checks. Or all values are passed in the same function and there are the dimension\ array length checks.

Reply:
> Or all values are passed in the same function and there are the dimension\ array length checks.

Agreed. Eg., 

```
void GeometryData::SetGeometryData(const IntegrationPointsContainerType& ThisIntegrationPoints,
                  const ShapeFunctionsValuesContainerType& ThisShapeFunctionsValues,
                  const ShapeFunctionsLocalGradientsContainerType& ThisShapeFunctionsLocalGradients );
```